src/workflow/CommandExecutor.py

```diff
@@ -0,0 +1,193 @@
+        
+        # Format the logging prefix
+        self.logger.log(f"Total time to run command: {execution_time:.2f} seconds")
+        
+        # Log stdout if present
+        if stdout and write_log:
+            self.logger.log(f"Console log:\n\n{stdout.decode()}")
+        
+        # Log stderr and raise an exception if errors occurred
+        if stderr or process.returncode != 0:
+            error_message = stderr.decode().strip()
+            self.logger.log(f"ERRORS OCCURRED:\n{error_message}")
+            raise Exception(f"Errors occurred while running command: {' '.join(command)}\n{error_message}")
+
+    def run_topp(self, tool: str, input_output: dict, write_log: bool = True) -> None:
+        """
+        Constructs and executes commands for the specified tool OpenMS TOPP tool based on the given
+        input and output configurations. Ensures that all input/output file lists
+        are of the same length, or single strings, to maintain consistency in command
+        execution. Supports executing commands either as single or multiple processes
+        based on the input size.
+
+        Args:
+            tool (str): The executable name or path of the tool.
+            input_output (dict): A dictionary specifying the input and output
+                                 parameters and their corresponding files. The files
+                                 can be specified as single paths (strings) or lists
+                                 of paths for batch processing.
+            write_log (bool): If True, enables logging of command execution details.
+        
+        Raises:
+            ValueError: If the lengths of input/output file lists are inconsistent,
+                        except for single string inputs.
+        """
+        # check input: any input lists must be same length, other items can be a single string
+        # e.g. input_mzML : [list of n mzML files], output_featureXML : [list of n featureXML files], input_database : database.tsv
+        io_lengths = [len(v) for v in input_output.values() if len(v) > 1]
+
+        if len(set(io_lengths)) > 1:
+            raise ValueError(f"ERROR in {tool} input/output.\nFile list lengths must be 1 and/or the same. They are {io_lengths}.")
+
+        if len(io_lengths) == 0:  # all inputs/outputs are length == 1
+            n_processes = 1
+        else:
+            n_processes = max(io_lengths)
+
+        commands = []
+
+        # Load parameters for non-defaults
+        params = ParameterManager().load_parameters()
+        # Construct commands for each process
+        for i in range(n_processes):
+            command = [tool]
+            # Add input/output files
+            for k in input_output.keys():
+                command += [f"-{k}"]
+                value = input_output[k]
+                if isinstance(value, Files):
+                    value = value.files
+                if len(value) == 1:
+                    i = 0
```

Comment:
This is a bit confusing to me. Could you add a comment on why this is needed? As far as I understood, this is for the case when multiple input/output files exist (e.g., multiple mzMLs and featureXMLs), but only one additional input file (e.g., one input database file) exists? But, in that case, the error will arise on line 139, will it not?

Reply:
Added some comments to clarify. Like you explained, when there are multiple files and also single files which should be in all runs (e.g. database file) only the first (and only) with index 0 should be picked for the single file.

```
  # get value from input_output dictionary
  value = input_output[k]
  # when multiple input/output files exist (e.g., multiple mzMLs and featureXMLs), but only one additional input file (e.g., one input database file)
  if len(value) == 1:
      i = 0
  # when the entry is a list of collected files to be passed as one [["sample1", "sample2"]]
  if isinstance(value[i], list):
      command += value[i]
  # standard case, files was a list of strings, take the file name at index
  else:
      command += [value[i]]
```

Line 139 checks if the inputs/output lengths are compatible. E.g. we have 3 mzML (input), 3 featureXML (output) and 1 database file it works. If there would be not the same number of input and output files it's not possible to construct a command.

Reply:
Adding "only one additional input file" comment clears things up for me! Thanks